packages/babel-helpers/src/helpers.js

```diff
@@ -814,3 +814,125 @@ helpers.applyDecoratedDescriptor = defineHelper(`
         return desc;
     }
 `);
+
+helpers.wrapRegExp = defineHelper(`
+  var kGroups = Symbol("@@babel/groups");
+
+  function BabelRegExp(re, groups) {
+    var _this = Reflect.construct(RegExp, [re], BabelRegExp);
+    _this[kGroups] = groups;
+    return _this;
+  }
+  BabelRegExp.prototype = Object.create(RegExp.prototype);
+  BabelRegExp.prototype.exec = function(str) {
+    var result = RegExp.prototype.exec.call(this, str);
+    if (result) result.groups = buildGroups(result, this);
+    return result;
+  };
+  BabelRegExp.prototype[Symbol.replace] = function(str, substitution) {
+    if (typeof substitution === "string") {
+      var groups = this[kGroups];
+      return RegExp.prototype[Symbol.replace].call(
+        this,
+        str,
+        substitution.replace(
+          /\\$<([^>]+)>/g,
+          function(_, name) {
+            return "$" + groups[name];
+          }
+        )
+      );
+    } else if (typeof substitution === "function") {
+      var _this = this;
+      return RegExp.prototype[Symbol.replace].call(
+        this,
+        str,
+        function() {
+          var args = [];
+          args.push.apply(args, arguments);
+          args.push(buildGroups(args, _this));
+          return substitution.apply(this, args);
+        }
+      );
+    } else {
+      return RegExp.prototype[Symbol.replace].call(this, str, substitution);
+    }
+  }
+
+  function buildGroups(result, re) {
+    return Object.keys(re[kGroups]).reduce(function(groups, name) {
+      groups[name] = result[re[kGroups][name]];
+      return groups;
+    }, Object.create(null));
+  }
+
+  export default function _wrapRegExp(re, groups) {
+    return new BabelRegExp(re, groups);
+  }
+`);
+
+// Since this helper is very long, use short variable names:
+// - n: name
+// - r: result
+// - o: object
+// - v: value
+// - g: groups
+// - p: replacement
+helpers.wrapRegExpCompat = defineHelper(`
+  function define(o, n, v) {
+    return Object.defineProperty(o, n, { value: v })
+  }
+
+  function wrap(o, n, fn) {
+    var f = o[n];
+    if (f["@@babel/wrapped"]) return;
+    define(o, n, function (x1, x2, x3) {
+      return fn.call(this, f, x1, x2, x3)
+    });
+    define(o[n], "@@babel/wrapped", true);
+  }
+
+  function groups(r, re) {
+    return Object.keys(re["@@babel/groups"]).reduce((g, n) => {
+      g[n] = r[re["@@babel/groups"][n]];
+      return g;
+    }, Object.create(null));
+  }
+
+  wrap(RegExp.prototype, "exec", function(f, str) {
```

Comment:
This is effectful, I understand that it's hard (impossible?) to implement this feature without actually messing with built in prototypes - but when that is the case I feel like it should not be a part of babel-helpers, but rather babel-polyfill

Reply:
@zloirock Does corejs polyfill support for `[String.replace]` and overwritten `exec`? If it does, we can remove thi "compat" helper from Babel.

Reply:
@nicolo-ribaudo `core-js` supports `@@replace`, but does not overwrite `exec`.

Reply:
Sorry, I didn't explain myself correctly. Every RegExp method (or string method which uses regexs) should use `exec` internally. e.g.

```js
class MyRegExp extends RegExp {
  exec() {
    return ["foo", "bar"];
  }
}

"abc".match(new MyRegExp("b")); // ["foo", "bar"]
```

If it doesn't work in `core-js`, I could open a PR

Reply:
`core-js` provides a naive realisation of support `RegExp`-related well-known symbols and use original methods like `String.prototype.match` or `String.prototype.replace`. It's done for reducing slowdown in wrappers. However, now it should not cause too many problems, so feel free to add a PR with adding this logic.